feat(mds,chunkserver): Add CLUSTER_ID config var

lgsilva3087 opened the pull request:
This change adds the CLUSTER_ID configuration variable for metadata (master, shadow) and chunkservers. The new variable should be set at the beginning and it is not intended to be changed frequently.

Metadata servers only accept connections from chunkservers with the same CLUSTER_ID (sent during registration), this prevents connections from undesired chunkservers.

To avoid breaking existent installations, the new variable defaults to 'default'. It is advisable for new installations to provide a meaningfull CLUSTER_ID.

lgsilva3087 committed:
feat(mds,chunkserver): Add CLUSTER_ID config var

This change adds the CLUSTER_ID configuration variable for metadata
(master, shadow) and chunkservers. The new variable should be set at
the beginning and it is not intended to be changed frequently.

Metadata servers only accept connections from chunkservers with the
same CLUSTER_ID (sent during registration), this prevents connections
from undesired chunkservers.

The variable defaults to `default`, to not break compatibility with
existing installations.

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 commented on src/master/matocsserv.cc:
```diff
@@ -964,35 +964,58 @@ void matocsserv_got_duptruncchunk_status(matocsserventry* eptr, const std::vecto
+			safs::log_err("SAU_CSTOMA_REGISTER_HOST received empty cluster ID");
+			eptr->mode = KILL;
+			return;
+		}
```
Done.

lgsilva3087 commented on src/master/matocsserv.cc:
```diff
@@ -964,35 +964,58 @@ void matocsserv_got_duptruncchunk_status(matocsserventry* eptr, const std::vecto
+		if (clusterId.empty() || clusterId != gClusterId) {
+			safs::log_err("SAU_CSTOMA_REGISTER_HOST received empty cluster ID");
+			eptr->mode = KILL;
+			return;
```
Done.

lgsilva3087 commented on src/chunkserver/master_connection.cc:
```diff
@@ -43,6 +43,7 @@
 #include "common/output_packet.h"
 #include "common/sockets.h"
 #include "config/cfg.h"
+#include "errors/saunafs_error_codes.h"
```
Removed.

lgsilva3087 commented on src/chunkserver/masterconn.cc:
```diff
@@ -262,14 +262,21 @@ int masterconn_init(void) {
 	gTimeout_ms = get_cfg_timeout();
 	gEnableLoadFactor = static_cast<bool>(cfg_getuint32("ENABLE_LOAD_FACTOR", 0));
 
+	if (clusterId.empty()) {
+		safs::log_err("masterconn_init: CLUSTER_ID is not set or empty in the configuration");
+		return -1;
+	}
```
The specific error is logged.

lgsilva3087 commented on src/master/filesystem.cc:
```diff
@@ -338,6 +338,12 @@ static void fs_read_goal_config_file() {
+
+	if (gClusterId.empty()) {
+		throw ConfigurationException("CLUSTER_ID is not defined in configuration file");
+	}
```
The text of the exception, which is printed, is clarifying the error.

lgsilva3087 commented on src/master/matocsserv.cc:
```diff
@@ -1024,13 +1047,23 @@ void matocsserv_space(matocsserventry *eptr,const uint8_t *data,uint32_t length)
+	deserializePacketVersionNoHeader(data, packetVersion);
+
+	if (packetVersion == cstoma::registerHost::kWithClusterId) {
+		cstoma::registerHost::deserialize(data, servip, servport, timeout, version, clusterId);
```
Older chunkservers can not use this version of the protocol packet.

lgsilva3087 commented on tests/tools/saunafs.sh:
```diff
@@ -332,6 +332,7 @@ create_bdb_name_storage_entry_() {
 create_sfsmaster_master_cfg_() {
 	local this_module_cfg_variable="MASTER_${masterserver_id}_EXTRA_CONFIG"
 	echo "PERSONALITY = master"
+	echo "CLUSTER_ID = cluster_01"
```
This is a default value, it can be overwritten by using the EXTRA_CONFIG family of variables.

lgsilva3087 committed:
chore(chunkserver): Improve MasterConn logs

Most of the logs now include the master's ip and port.

Signed-off-by: guillex <guillex@leil.io>

uristdwarf commented on src/master/filesystem.cc:
```diff
@@ -338,6 +338,12 @@ static void fs_read_goal_config_file() {
 }
 
 static void fs_read_config_file() {
+	gClusterId = cfg_getstring("CLUSTER_ID", "");
+
+	if (gClusterId.empty()) {
+		throw ConfigurationException("CLUSTER_ID is not defined in configuration file");
+	}
+
```
Main problem I see here is that existing setups could stop working unexpectedly, thus making this a breaking change

uristdwarf replied:
Maybe it's better that the program itself generates a UUID and then stores it in /var/lib/saunafs (or whatever the data directory is) in a file? Because this also potentially introduces user error with defining the config variables.

lgsilva3087 replied:
Nice catchs. I added a default value for this option, which will not break existing installations. Anyway, new installations could start using it to protect from connections from non intended chunkservers.
I agree, this function is expected to evolve in the future to be more robust (when multiple masters will be allowed), this change just adds it to the protocol.

uristdwarf replied:
Just be sure to remove the breaking ! symbol from the commit

uristdwarf replied:
As that can cause confusion when the changelog is generated

uristdwarf replied:
I guess my only final concern is that are you sure you don't want to change this configuration option later (e.g not have it as a configuration option or something similar)? If not, you might consider making the option experimental (even though it's already used). Just a thought.

Anyway, feel free to close this thread after you've read my messages

ralcolea commented on src/master/matocsserv.cc:
```diff
@@ -964,35 +964,66 @@ void matocsserv_got_duptruncchunk_status(matocsserventry* eptr, const std::vecto
+	// Send the answer with the status
+	if (eptr->version >= kFirstVersionWithClusterId) {
+		OutputPacket outPacket;
+		matocs::registerHost::serialize(outPacket.packet, 0);
```
If **0** represents a successful registration, please consider using `SAUNAFS_STATUS_OK`.

lgsilva3087 replied:
Done.

ralcolea reviewed (approved):
Great job @lgsilva3087!!! :+1: :fire: :muscle:

dmga44 commented on src/chunkserver/master_connection.cc:
```diff
@@ -318,15 +347,16 @@ void MasterConn::readFromSocket() {
-				mode_ = ConnectionMode::KILL;
+				safs::log_error_code(errno, "MasterConn: read from Master error: {}",
+				                     address_.toString());
+				setMode(ConnectionMode::KILL);
```
Please note here that it will print something like `MasterConn: read from Master error: 192.168.50.206 (Read error)`.

lgsilva3087 replied:
Done.

dmga44 commented on src/protocol/cstoma.h:
```diff
@@ -50,13 +50,24 @@ SAUNAFS_DEFINE_PACKET_SERIALIZATION(
+
 SAUNAFS_DEFINE_PACKET_SERIALIZATION(
-		cstoma, registerHost, SAU_CSTOMA_REGISTER_HOST, 0,
+		cstoma, registerHost, SAU_CSTOMA_REGISTER_HOST, kBeforeClusterId,
```
Maybe `kWithoutClusterId` would look better?

lgsilva3087 replied:
Done.

rolysr reviewed (approved):
LGTM.

dmga44 commented on src/chunkserver/master_connection.cc:
```diff
@@ -98,7 +98,7 @@ void MasterConn::reloadConfig() {
 		}
 	} else {
-		safs::log_warn("master connection module: can't resolve master host/port ({}:{})",
+		safs::log_warn("MasterConn: can't resolve master host/port ({}:{})",
```
On the `MasterConn::reloadConfig` function: would it make sense to also reload the `CLUSTER_ID` parameter? Please consider it.

If not, it would make sense to add a message that it is not being reloaded.

lgsilva3087 replied:
The CLUSTER_ID should be set at the beginning of the installation and it is not expected to change frequently, just in very specific rare situations. I added the message to warn if the user tries to reload it.

dmga44 reviewed (approved):
LGTM.

Taking into consideration this is a breaking change, that our major release is almost done, does it make sense to merge it to dev?

lgsilva3087 committed:
feat(mds,chunkserver): Send Master info at registration

During the registration, Master server now answers with an status
code, its version and the CLUSTER_ID. After registration, the
chunkservers can now use better the protocol versioning to talk to an
specific Master version.

Signed-off-by: guillex <guillex@leil.io>
